### ISSUE TITLE
command failure result in queue clearing

### ISSUE THREAD

**Issue by ledbit** [steveukx/git-js#401]
Experiencing an issue where a git command failure clears up the queued commands after it - there are certain, exploratory commands where a failure is OK. This can be easily reproduced by sending in two requests, the first of which fails and you'd notice the second one never to resolve

```
const Git = require('simple-git/promise');

const git = Git(__dirname);
const p = git.raw(['remote', 'get-url', 'notThere']).catch(()=>{});
const p1 = git.log();
p1.then(() => {console.log('it will never get here');})

``` 

The root cause seems to be this line - before sending a PR I want to understand if there were any reasons for clearing up the queue or was just a left over from .exception?
https://github.com/steveukx/git-js/blob/master/src/git.js#L1472

**Comment by Elecweb** [comment]
I also have this issue. is there any workaround?

**Comment by steveukx** [comment]
As chained commands are treated as a single workflow, errors intentionally prevent the remaining chained steps from executing. If the steps are independent of each other (ie: you can do the second step whether the first fails or not), you can write them as:

```
const gitP = require('simple-git/promise');
const git = gitP();

git.raw([...]).catch(console.error) // runs git.raw, catches an exception if one occurs
  .then(() => git.log()) // then runs git.log
  .then(console.log) // and logs that to the console
```

**Comment by alisey** [comment]
@steveukx in event-driven programs the commands can run independently of each other, and it wouldn't be possible to chain them like you suggest.

It would be helpful to have a disclaimer in the readme explaining the implications of treating commands as a single workflow. It's a surprising behavior that might affect the decision about whether Simple Git is suitable for certain projects.

**Comment by alisey** [comment]
Command failure no longer leaves pending commands in an unresolved state indefinitely, but it still isn't possible to run independent commands in parallel using a single instance of SimpleGit.

As a workaround I create a new SimpleGit instance for each command:
```
const simpleGit = require('simple-git');

await Promise.all([
    simpleGit().raw('X'),
    simpleGit().raw('Y')
]);
```

@steveukx, would you consider one of the following PRs:

1) `.disableQueue()` method that disables queue for all future tasks in a given instance of SimpleGit
2) `.startQueue()` method that internally replaces `SimpleGit._executor` with a new instance of `GitExecutor`
3) enable or disable queue depending on whether method calls are chained or not:
    - `git.raw('X').raw('Y')` - run X then Y
    - `git.raw('X'); git.raw('Y');` - run X and Y in parallel

In the option 3 the `_runTask()` method instead of `this` would return a clone of `this` with an extra internal flag that enables chaining.

**Comment by steveukx** [comment]
@alisey if you update to version `2.11.0` you should now see that chained tasks work atomically (where one fails, the whole promise chain will reject), but each task created from the main `simpleGit` instance is the start of a new chain, and will be executed whether other chains fail or not.

Some time ago (see #317) it was shown that spawning too many `git` child processes could mean their responses were lost to `node`. To prevent this, each `simpleGit` instance limits the number of child processes it has open at any one time, if necessary this can be changed by using the new [configuration option](https://github.com/steveukx/git-js/tree/v2.11.0#configuration) `{ maxConcurrentProcesses: 5 }`

The documentation has been updated in:

- https://github.com/steveukx/git-js/tree/v2.11.0#concurrent--parallel-requests
- https://github.com/steveukx/git-js/tree/v2.11.0#configuration

**Comment by alisey** [comment]
@steveukx I've updated and it's working great. Thank you so much! Also tested various values for `maxConcurrentProcesses`, which is a really nice addition, also works as expected.

**Comment by steveukx** [comment]
I am going to close this issue now - if there further problems, please do open a new issue.
